JASP Encryption Files

RensDofferhoff opened the pull request:
JASP encryption files implemented using libsodium asym encryption.
A password is used to seed keypair generation.
 
Algo details: https://doc.libsodium.org/public-key_cryptography/authenticated_encryption
This stuff is top notch but the asymmetric case is  not quantum safe (yet,  its on the libsodium roadmap but realistically not an issue for looooong time).
symetric case is  quantum safe

Can be used to encrypt JASP files with a password.
Optional JASP submission checkbox so the JASP-Team can also decrypt the files in question.

How to use:

On the Save file select window, select Encrypted JASP file as type.
<img width="1680" height="1210" alt="image" src="https://github.com/user-attachments/assets/eac24fd1-d879-4894-82de-2c516e9a4310" />

Enter Encryption details:
<img width="744" height="416" alt="image" src="https://github.com/user-attachments/assets/18dd8323-1e61-48db-99f2-594cf145a036" />


Chosen password hash function is a bit hefty to make weak password rainbow tables expensive.

RensDofferhoff committed:
Encryption Initial

fix compilation mac

make encrypt windows a bit nicer

cleanup

make symetric case quantum safe for free

add support for custom public/private keys

Update ubuntu.yml

Update ubuntu.yml

Update dialog

RensDofferhoff commented:
wish to add one more thing

shun2wang commented:
Build test on Linux was failed.

boutinb commented:
Great feature!!
I have updated a bit the encryption dialog box (by using a Group, you get automatically alignment of the TextFields).
I found the feature a bit hidden when you save a JASP. It is (nearly) impossible to add a checkbox to a native dialog (default dialogs are native), so I understand that you chose to add a filter (even if it has the same jasp suffix), but I'm afraid that most of the users won't see this option. That's why I have changed the title of the dialog by 'Save (encrypted) JASP file". But maybe someone has a better idea to increase the attention of the user for this feature.

2 issues:

- If you do Save As with encryption, if you close the encryption dialog, it still saves the file.
- if you load an encrypted file, if you close the encryption dialog, it still tries to load the file, and you get an error that it could not load it: in this case it should not try to load it.

shun2wang commented:
>. But maybe someone has a better idea to increase the attention of the user for this feature.


We can add sub menu while cursor hover the "save as" button then push two options to the right, a "save as", and a "save as 
encrypted",and the second option always comes with an encryption dialog.

RensDofferhoff commented:
That's a good suggestion! I will implement this next week if Bruno agrees. But most people use ctrl+s I fear not the menu

Sent from Proton Mail Android

-------- Original Message --------
On 10/3/25 13:54, Shun Wang  wrote:

> shun2wang left a comment [(jasp-stats/jasp-desktop#6029)](https://github.com/jasp-stats/jasp-desktop/pull/6029#issuecomment-3365402565)
>
>> . But maybe someone has a better idea to increase the attention of the user for this feature.
>
> We can add sub menu while cursor hover the "save as" button then push two options to the right, a "save as", and a "save as
> encrypted",and the second option always comes with an encryption dialog.
>
> —
> Reply to this email directly, [view it on GitHub](https://github.com/jasp-stats/jasp-desktop/pull/6029#issuecomment-3365402565), or [unsubscribe](https://github.com/notifications/unsubscribe-auth/AFABXSYQW2WKB6EWC7BU3JL3VZPYLAVCNFSM6AAAAACHMOSQ52VHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZTGNRVGQYDENJWGU).
> You are receiving this because you authored the thread.Message ID: ***@***.***>

boutinb commented:
> That's a good suggestion! I will implement this next week if Bruno agrees. But most people use ctrl+s I fear not the menu 

I don't see exactly what you mean: there is now already 2 submenus (Computer and OSF). Do you want to add an extra submenu 'Computer with encryption' (I don't think that encryption can be added for OSF) ? Better is to add a checkbox 'Use encryption' in the Computer submenu. That would set de JASP encryption fiter per default in the dialog.
But as you said, this won't work with the CTRL-S shortcut. So I think to set in the title ''Save (encrypted) JASP file" is still needed.

shun2wang commented:
Ctrl+S shortcut is works for a saved file (whatever it is Encrypted or not) but should not a "save as...", in another case if user pressed that key from a created new file then it should just popup a save dialog ask save path. for example in Excel if you created a file from "mew file" button when you use shotcut to save it, application will ask you where to save it.

JorisGoosen commented:
> Great feature!! I have updated a bit the encryption dialog box (by using a Group, you get automatically alignment of the TextFields). I found the feature a bit hidden when you save a JASP. It is (nearly) impossible to add a checkbox to a native dialog (default dialogs are native), so I understand that you chose to add a filter (even if it has the same jasp suffix), but I'm afraid that most of the users won't see this option. That's why I have changed the title of the dialog by 'Save (encrypted) JASP file". But maybe someone has a better idea to increase the attention of the user for this feature.

We could just not use the native dialog for this particular dialog.

Is it then easier to add a checkbox perhaps?

JorisGoosen commented:
Works pretty smooth, however if you cancel the save dialog before entering the password and all the progressbar gets stuck. So I'll fix that.

JorisGoosen committed:
Make sure the progressbar disappears if no password is selected during saving

also make sure that if the user doesnt click "submit" there it doesnt just save it with an empty password, but instead just cancels the save.

JorisGoosen commented:
Alright, if looked through the code, it seems fine to me.
Ive made sure the progressbar doesnt get stuck if you fail to submit a password.

If @boutinb also thinks this is fine we could perhaps merge this?
I didnt test on windows or linux yet though.

JorisGoosen commented on QMLComponents/utilities/desktopcommunicator.cpp:
```diff
@@ -34,4 +38,26 @@ bool DesktopCommunicator::engineSandbox()
+
+	queryCondition = false;
+	std::unique_lock<std::mutex> lock(queryLock);
+	emit queryEncryptionSettingsSignal();
```
This all couldve just been done with `Qt::BlockingQueuedConnection` but I cant be arsed to rewrite it for that now.

boutinb committed:
Remove JASP Team Submission CHeckBox when reading

boutinb committed:
Make the Encryption setting window resizable and handle cancel button

boutinb committed:
Use exception instead of boolean.

Also fix auto save with encrypted file

boutinb reviewed (approved):
I think we can merge this now, @JorisGoosen ?
